### ISSUE TITLE
Feature request: support for environment variables in configuration files

### ISSUE THREAD

**Issue by jmhostalet** [eclipse/mosquitto#2203]
Given a configuration files, for example:
```
/etc/mosquitto/conf.d/my-bridge.conf
```
it would be nice to use environment variables, for example:
```
bridge_identity my-identity
bridge_psk ${MOSQUITTO_PSK}
```
This would help a lot to use the docker images without having to generate a custom image by extending the official one.
It would also help a lot in the use of kubernetes, making use of the resources of type "secret" and mounting them in the container itself as environment variables.

**Comment by ralight** [comment]
This seems like a good idea, I'm now wondering about what vulnerabilities this could introduce and how to mitigate them.

**Comment by sglux** [comment]
Don't know if this introduces weaknesses but maybe the requested "feature" can be turned on and of with a switch parameter? If the hypothetical switch `--use-env` is not specified nothing is changed to the interpretation of configuration files.

**Comment by onedr0p** [comment]
> I'm now wondering about what vulnerabilities this could introduce and how to mitigate them.

What did you have in mind? This is a pretty common design pattern and follows the [12factor app](https://12factor.net/).

**Comment by beckerben** [comment]
Doesn't seem to be a issue for other brokers, my favorite is vernemq, easy to run as docker container because all settings can be configured from environment variables...also hive-ce supports environment variable tokens embedded in the file which is nice too.  There is a version of mosquitto broker actually that someone made to support the variables, would be nice if this was baked into official build: https://github.com/efrecon/docker-mosquitto

**Comment by abiliojr** [comment]
I was half way into a prototype for this, when it struck me: wouldn't it be enough to use something like envsubst for the docker image. Is support wanted outside of Docker that useful?

**Comment by beckerben** [comment]
@abiliojr that is a good solution for building an image, however the issue is once you have an image built and now want to use it and change the configuration for each instance of a container where the common image is used, this method doesn't work.

**Comment by abiliojr** [comment]
@beckerben, that is true. But my ideas was to have it to be part of the docker container itself, running before mosquitto runs (so mosquitto consumes it's output).

**Comment by beckerben** [comment]
@abiliojr interesting,  would the base image of the image be a linux distro like alpine you install mosquitto on top of and use a script in the image to build the configuration and then start mosquitto or is there another method?
